File: frontend/types/assessment.ts
export type AssessmentStatus = 'draft' | 'in_progress' | 'review' | 'completed' | 'abandoned' | 'archived'
export type SecurityLevel = 'osnovna' | 'srednja' | 'napredna'

export interface AssessmentProgress {
  total_controls: number
  completed_controls: number
  mandatory_controls: number
  completed_mandatory: number
  completion_percentage: number
  sections_completed: number
  total_sections: number
  last_activity: string
  // Aliases used by V2/V3 APIs and UI components
  answered_controls?: number
  mandatory_answered?: number
}

export interface AssessmentScores {
  average_documentation_score: number
  average_implementation_score: number
  compliance_score: number
  mandatory_compliance_score: number
  voluntary_compliance_score: number
  by_measure: Record<string, {
    documentation_score: number
    implementation_score: number
    compliance_score: number
  }>
}

export interface Assessment {
  id: string
  organization_id: string
  title: string
  description?: string
  security_level: SecurityLevel
  status: AssessmentStatus
  progress?: AssessmentProgress
  current_scores?: AssessmentScores
  created_at: string
  updated_at: string
  created_by: string
  updated_by?: string
  assigned_to?: string[]
  submitted_at?: string
  submitted_by?: string
  due_date?: string
  tags?: string[]
  // Optional fields returned by list/detail APIs and used in UI
  total_controls?: number
  answered_controls?: number
  mandatory_controls?: number
  mandatory_answered?: number
  completion_percentage?: number
  mandatory_completion_percentage?: number
  compliance_percentage?: number
  completed_at?: string
}

export interface CreateAssessmentRequest {
  title: string
  description?: string
  organization_id: string
  security_level: SecurityLevel
  due_date?: string
  assigned_to?: string[]
}

export interface CreateAssessmentResponse {
  success: boolean
  assessment_id: string
  title: string
  status: string
  security_level: string
  created_at: string
}

export interface UpdateAssessmentRequest {
  title?: string
  description?: string
  status?: AssessmentStatus
  due_date?: string
  tags?: string[]
}

export interface AssessmentAnswer {
  control_id: string
  submeasure_id: string  // Required for submeasure-specific answers
  documentation_score?: number | null
  implementation_score?: number | null
  comments?: string | null
  evidence_files?: string[]
  confidence_level?: number | null
}

export interface SubmitAnswersRequest {
  answers: AssessmentAnswer[]
}

export interface ControlRatingGuidance {
  score: number
  documentation_criteria: string
  implementation_criteria: string
}

export interface AssessmentControl {
  id: string
  code: string
  name_hr: string
  description_hr?: string
  order_index: number
  is_mandatory: boolean  // Deprecated - use submeasure_context
  requirement_id: string
  security_level: string
  documentation_score: number | null
  implementation_score: number | null
  comments?: string | null
  evidence_files?: string[]
  answered_at?: string | null
  answered_by?: string | null
  // New fields for minimum score requirements
  minimum_score?: number | null
  submeasure_id?: string | null  // Deprecated - use submeasure_context
  submeasure_context?: {  // New: context for current submeasure
    submeasure_id: string
    is_mandatory: boolean
    minimum_score: number
  }
  meets_minimum?: boolean | null
  rating_guidance?: ControlRatingGuidance[]
  submeasure_ids?: string[]  // New: all submeasures this control belongs to
}

export interface AssessmentSubmeasure {
  id: string
  code: string
  name_hr: string
  description_hr?: string
  order_index: number
  controls: AssessmentControl[]
  // Aggregate scoring fields
  total_controls: number
  answered_controls: number
  mandatory_controls: number
  mandatory_answered: number
  documentation_avg?: number | null
  implementation_avg?: number | null
  overall_score?: number | null
  passes_threshold?: boolean | null
}

export interface AssessmentMeasure {
  id: string
  code: string
  name_hr: string
  description_hr?: string
  order_index: number
  submeasures: AssessmentSubmeasure[]
  // Aggregate fields
  total_controls: number
  answered_controls: number
  mandatory_controls: number
  mandatory_answered: number
  overall_score?: number | null
}

export interface AssessmentQuestionnaire {
  assessment_id: string
  security_level: string
  version_id: string
  measures: AssessmentMeasure[]
  statistics: {
    total_measures: number
    total_submeasures: number
    total_controls: number
    mandatory_controls: number
    voluntary_controls: number
  }
  generated_at: string
}

export interface MeasureResult {
  measure_id: string
  measure_title: string
  measure_code: string
  compliance_percentage: number
  avg_documentation_score?: number
  avg_implementation_score?: number
  total_controls: number
  answered_controls: number
  mandatory_total: number
  mandatory_completed: number
  submeasure_count: number
  control_count: number
  gap_areas?: string[]
}

export interface AssessmentResults {
  assessment_id: string
  overall_compliance_percentage: number
  mandatory_compliance_percentage: number
  voluntary_compliance_percentage: number
  documentation_average: number
  implementation_average: number
  total_controls: number
  answered_controls: number
  mandatory_total: number
  mandatory_completed: number
  measure_results?: MeasureResult[]
  by_measure: Record<string, {
    measure_code: string
    measure_name: string
    compliance_score: number
    documentation_score: number
    implementation_score: number
    total_controls: number
    completed_controls: number
  }>
  gaps: Array<{
    control_id: string
    control_code: string
    control_name: string
    gap_score: number
    is_mandatory: boolean
    current_documentation: number
    current_implementation: number
  }>
  recommendations: string[]
}

export interface GapAnalysisItem {
  id: string
  control_id: string
  control_title: string
  measure_title: string
  priority: 'critical' | 'high' | 'medium' | 'low'
  category: 'mandatory' | 'implementation' | 'documentation'
  current_score: number
  target_score: number
  gap_score: number
  effort_estimate: 'high' | 'medium' | 'low'
  timeline_weeks: number
  impact_description: string
  recommendations: string[]
}

export interface AIRecommendation {
  id: string
  title: string
  description: string
  priority: 'critical' | 'high' | 'medium' | 'low'
  effort_estimate: 'high' | 'medium' | 'low'
  timeline_weeks: number
  compliance_impact: number
  implementation_steps?: string[]
  source_references?: string[]
  category?: string
  control_ids?: string[]
  is_implemented?: boolean
  created_at?: string
  measure_name?: string
  control_name?: string
}

// Compliance Scoring Types
export interface SubmeasureResult {
  submeasure_id: string
  submeasure_code: string
  documentation_avg?: number
  implementation_avg?: number
  overall_score: number
  passes_individual_threshold: boolean
  passes_average_threshold: boolean
  passes_overall: boolean
  mandatory_controls_count?: number
  failed_controls: string[]
  control_count?: number  // Added for total controls in submeasure
  answered_count?: number  // Added for answered controls
  mandatory_count?: number  // Added for mandatory controls
  mandatory_answered?: number  // Added for answered mandatory controls
  // New fields from enhanced API
  total_controls?: number
  answered_controls?: number
  mandatory_controls?: number
  
}

export interface MeasureResultCompliance {
  measure_id: string
  measure_code: string
  overall_score: number
  documentation_avg?: number  // Added from API response
  implementation_avg?: number  // Added from API response
  compliance_percentage?: number  // Added from API response
  passes_compliance: boolean
  total_submeasures: number
  passed_submeasures: number
  critical_failures: string[]
  submeasures?: SubmeasureResult[]  // Added to support nested submeasures
  // Control count fields from MeasureScore model
  total_controls?: number  // Added from stored model
  answered_controls?: number  // Added from stored model
  mandatory_controls?: number  // Added from stored model
  mandatory_answered?: number  // Added from stored model
}

export interface ComplianceResult {
  overall_compliance_score: number
  compliance_percentage: number
  passes_compliance: boolean
  total_measures: number
  passed_measures: number
  maturity_score: number
  maturity_threshold: number
  meets_maturity_trend: boolean
  security_level: string
  thresholds: {
    individual: number
    average: number
  }
}

export interface AssessmentResultsResponse {
  assessment_id: string
  overall_score: number
  compliance_percentage: number
  measure_results: MeasureResultCompliance[]
  submeasure_results: SubmeasureResult[]
  calculated_at: string
  statistics: {
    total_controls: number
    answered_controls: number
    mandatory_controls: number
    mandatory_answered: number
    compliance: ComplianceResult
  }
}

// API Response Types - Backend returns nested structure
export interface AssessmentDetailResponse {
  assessment: Assessment
  scores: Record<string, any>
  progress: Record<string, any>
  validation: Record<string, any>
  active_users: number
  statistics: Record<string, any>
  valid_transitions: string[]
}

// V2 API Types with M:N support
export interface ControlInQuestionnaireResponseV2 {
  control: {
    id: string
    code: string
    name: string
    description: string
    guidance?: string
  }
  context: {
    submeasure_id: string
    is_mandatory: boolean
    minimum_score: number
  }
  answer?: {
    id: string
    documentation_score?: number
    implementation_score?: number
    overall_score?: number
    comments?: string
    evidence_files?: string[]
  }
}

export interface SubmeasureInQuestionnaireResponseV2 {
  id: string
  number: string
  name: string
  description?: string
  controls: ControlInQuestionnaireResponseV2[]
  statistics: {
    total_controls: number
    mandatory_controls: number
    answered_controls: number
    mandatory_answered: number
    mandatory_meeting_minimum: number
    compliance_percentage: number
    compliance_issues: string[]
  }
}

export interface MeasureInQuestionnaireResponseV2 {
  id: string
  number: string
  name: string
  description?: string
  submeasures: SubmeasureInQuestionnaireResponseV2[]
  statistics: {
    total_controls: number
    answered_controls: number
    compliance_percentage: number
  }
}

export interface QuestionnaireResponseV2 {
  assessment_id: string
  security_level: string
  measures: MeasureInQuestionnaireResponseV2[]
  statistics: {
    total_controls: number
    unique_controls: number  // Since controls can appear in multiple submeasures
    answered_controls: number
    completion_percentage: number
    compliance_score: number
    mandatory_controls: number
    mandatory_answered: number
  }
}

// V3 API Types
export interface ComplianceStatus {
  assessment_id: string
  security_level: string
  status: 'compliant' | 'non_compliant' | 'in_progress'
  passed_submeasures: number
  total_submeasures: number
  passed_measures: number
  total_measures: number
  average_score: number
  minimum_score: number
  target_average: number
  compliance_percentage: number
  maturity_score: number
  maturity_threshold: number
  details: {
    measure_compliance: {
      measure_id: string
      measure_code: string
      measure_name: string
      passed_submeasures: number
      total_submeasures: number
      is_compliant: boolean
    }[]
    submeasure_compliance: {
      submeasure_id: string
      submeasure_code: string
      submeasure_name: string
      average_score: number
      all_controls_pass: boolean
      is_compliant: boolean
      control_details: {
        control_id: string
        control_code: string
        score: number
        minimum_required: number
        passes: boolean
      }[]
    }[]
  }
}

export interface ValidationResult {
  is_valid: boolean
  errors: string[]
  warnings: string[]
  mandatory_complete: boolean
  missing_mandatory: {
    control_id: string
    control_code: string
    control_name: string
    submeasure_code: string
  }[]
  statistics: {
    total_controls: number
    answered_controls: number
    mandatory_controls: number
    mandatory_answered: number
  }
}

export interface AssessmentInsightsGapItem {
  control_id: string
  control_code: string
  control_name: string
  priority: 'critical' | 'high' | 'medium' | 'low'
  is_mandatory: boolean
  current_score?: number | null
  target_score?: number | null
  recommendation?: string | null
}

export interface AssessmentInsightsRoadmapItem {
  control_id: string
  control_name: string
  priority: 'critical' | 'high' | 'medium' | 'low'
  recommendation?: string | null
}

export interface AssessmentInsightsRoadmapPhase {
  name: string
  duration: string
  description: string
  items: AssessmentInsightsRoadmapItem[]
}

export interface AssessmentInsights {
  assessment_id: string
  computed_at?: string | null
  gaps: AssessmentInsightsGapItem[]
  roadmap: {
    summary: string
    total_items: number
    phases: AssessmentInsightsRoadmapPhase[]
  }
  ai_summary?: string | null
  measures_ai: {
    measures: Array<{
      code: string
      name: string
      ai_summary?: string
      controls: Array<{
        control_id: string
        control_code: string
        control_name: string
        ai_recommendation?: string | null
      }>
    }>
  }
  status: 'ok' | 'stale' | 'error'
  source_version: string
}

File: backend/app/schemas/assessment.py
"""Pydantic schemas for assessment API endpoints."""

from datetime import datetime
from typing import Dict, List, Optional, Any
from uuid import UUID

from pydantic import BaseModel, Field


# Base schemas
class AssessmentAnswerResponse(BaseModel):
    """Assessment answer response schema."""

    id: UUID
    control_id: UUID
    documentation_score: Optional[int] = None
    implementation_score: Optional[int] = None
    comments: Optional[str] = None
    evidence_files: List[str] = Field(default_factory=list)
    confidence_level: Optional[int] = None
    answered_by: Optional[UUID] = None
    answered_at: Optional[datetime] = None
    is_final: bool = False

    model_config = {"from_attributes": True}


class AssessmentResponse(BaseModel):
    """Assessment response schema."""

    id: UUID
    title: str
    description: Optional[str] = None
    organization_id: UUID
    security_level: str
    status: str
    created_by: Optional[UUID] = None
    updated_by: Optional[str] = None  # User name of last person who updated the assessment
    assigned_to: Optional[List[UUID]] = None
    created_at: datetime
    updated_at: datetime
    started_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None
    due_date: Optional[datetime] = None

    # Scoring fields
    compliance_percentage: Optional[float] = None
    total_controls: Optional[int] = None
    answered_controls: Optional[int] = None
    mandatory_controls: Optional[int] = None
    mandatory_answered: Optional[int] = None

    model_config = {"from_attributes": True}


class AssessmentDetailResponse(BaseModel):
    """Detailed assessment response with additional context."""

    assessment: AssessmentResponse
    scores: Dict[str, Any] = Field(default_factory=dict)
    progress: Dict[str, Any] = Field(default_factory=dict)
    validation: Dict[str, Any] = Field(default_factory=dict)
    active_users: int = 0
    statistics: Dict[str, Any] = Field(default_factory=dict)
    valid_transitions: List[str] = Field(default_factory=list)


class AssessmentListResponse(BaseModel):
    """Assessment list response with pagination."""

    assessments: List[AssessmentResponse] = Field(default_factory=list)
    total: int = Field(description="Total number of assessments matching the query")
    limit: int = Field(description="Maximum number of results per page")
    offset: int = Field(description="Number of results skipped")
    has_more: bool = Field(description="Whether there are more results available")


class AssessmentResultResponse(BaseModel):
    """Assessment results response."""

    assessment_id: UUID
    overall_score: Optional[float] = None
    compliance_percentage: Optional[float] = None
    measure_results: List[Dict[str, Any]] = Field(default_factory=list)
    submeasure_results: List[Dict[str, Any]] = Field(default_factory=list)
    calculated_at: datetime
    statistics: Dict[str, Any] = Field(default_factory=dict)


class AssessmentProgressResponse(BaseModel):
    """Assessment progress response."""

    assessment_id: UUID
    completion_percentage: float = 0.0
    mandatory_completion_percentage: float = 0.0
    total_controls: int = 0
    answered_controls: int = 0
    mandatory_controls: int = 0
    mandatory_answered: int = 0
    measure_progress: List[Dict[str, Any]] = Field(default_factory=list)
    updated_at: datetime


class AssessmentValidationResponse(BaseModel):
    """Assessment validation response."""

    assessment_id: UUID
    valid: bool
    can_submit: bool
    missing_mandatory: List[UUID] = Field(default_factory=list)
    low_scores_without_comments: List[UUID] = Field(default_factory=list)
    validation_errors: List[str] = Field(default_factory=list)
    statistics: Dict[str, Any] = Field(default_factory=dict)


class AssessmentActivityResponse(BaseModel):
    """Assessment activity response."""

    assessment_id: UUID
    active_users: List[Dict[str, Any]] = Field(default_factory=list)
    recent_changes: List[Dict[str, Any]] = Field(default_factory=list)
    last_updated: Optional[datetime] = None


# Control response schema for questionnaire
class ControlInQuestionnaireResponse(BaseModel):
    """Control data within questionnaire structure."""
    
    id: UUID
    code: str
    name_hr: str
    description_hr: Optional[str] = None
    order_index: int
    is_mandatory: bool
    requirement_id: UUID
    security_level: str
    # Minimum score requirement
    minimum_score: Optional[float] = None
    submeasure_id: Optional[UUID] = None
    # Answer data if exists
    documentation_score: Optional[int] = None
    implementation_score: Optional[int] = None
    comments: Optional[str] = None
    evidence_files: List[str] = Field(default_factory=list)
    answered_at: Optional[datetime] = None
    answered_by: Optional[UUID] = None


# Submeasure response schema for questionnaire
class SubmeasureInQuestionnaireResponse(BaseModel):
    """Submeasure data within questionnaire structure."""
    
    id: UUID
    code: str
    name_hr: str
    description_hr: Optional[str] = None
    order_index: int
    controls: List[ControlInQuestionnaireResponse]
    # Aggregate scoring fields
    total_controls: int = 0
    answered_controls: int = 0
    mandatory_controls: int = 0
    mandatory_answered: int = 0
    documentation_avg: Optional[float] = None
    implementation_avg: Optional[float] = None
    overall_score: Optional[float] = None
    passes_threshold: Optional[bool] = None


# Measure response schema for questionnaire
class MeasureInQuestionnaireResponse(BaseModel):
    """Measure data within questionnaire structure."""
    
    id: UUID
    code: str
    name_hr: str
    description_hr: Optional[str] = None
    order_index: int
    submeasures: List[SubmeasureInQuestionnaireResponse]
    # Aggregate fields
    total_controls: int = 0
    answered_controls: int = 0
    mandatory_controls: int = 0
    mandatory_answered: int = 0
    overall_score: Optional[float] = None


class QuestionnaireResponse(BaseModel):
    """Questionnaire structure response for assessment."""

    assessment_id: UUID
    security_level: str
    version_id: UUID
    measures: List[MeasureInQuestionnaireResponse]
    statistics: Dict[str, Any] = Field(default_factory=dict)
    generated_at: datetime


# Request schemas
class CreateAssessmentRequest(BaseModel):
    """Create assessment request schema."""

    title: str = Field(min_length=1, max_length=255)
    description: Optional[str] = Field(None, max_length=1000)
    organization_id: UUID
    security_level: str = Field(pattern="^(osnovna|srednja|napredna)$")
    due_date: Optional[datetime] = None
    assigned_to: Optional[List[UUID]] = None


class UpdateAssessmentRequest(BaseModel):
    """Update assessment request schema."""

    title: Optional[str] = Field(None, min_length=1, max_length=255)
    description: Optional[str] = Field(None, max_length=1000)
    due_date: Optional[datetime] = None
    assigned_to: Optional[List[UUID]] = None


class UpdateAnswerRequest(BaseModel):
    """Update answer request schema."""

    control_id: UUID
    submeasure_id: UUID  # Required for submeasure-specific answers
    documentation_score: Optional[int] = Field(None, ge=1, le=5)
    implementation_score: Optional[int] = Field(None, ge=1, le=5)
    comments: Optional[str] = Field(None, max_length=2000)
    evidence_files: Optional[List[str]] = Field(default_factory=list)
    confidence_level: Optional[int] = Field(None, ge=1, le=5)


class BatchUpdateAnswersRequest(BaseModel):
    """Batch update answers request schema."""

    answers: List[UpdateAnswerRequest] = Field(min_length=1)


class TransitionStatusRequest(BaseModel):
    """Status transition request schema."""

    new_status: str = Field(
        pattern="^(draft|in_progress|review|completed|abandoned|archived)$"
    )
    reason: Optional[str] = None
    current_status: Optional[str] = None
    force: bool = False


class AssignUsersRequest(BaseModel):
    """Assign users request schema."""

    user_ids: List[UUID] = Field(min_length=1)


# Search and filter schemas
class AssessmentSearchRequest(BaseModel):
    """Assessment search request parameters."""

    search_term: Optional[str] = Field(
        None, description="Search term for assessment title/description"
    )
    organization_id: Optional[UUID] = Field(None, description="Filter by organization")
    status: Optional[str] = Field(None, description="Filter by status")
    security_level: Optional[str] = Field(None, description="Filter by security level")
    assigned_user_id: Optional[UUID] = Field(
        None, description="Filter by assigned user"
    )
    created_after: Optional[datetime] = Field(
        None, description="Filter by creation date (after)"
    )
    created_before: Optional[datetime] = Field(
        None, description="Filter by creation date (before)"
    )
    limit: int = Field(50, ge=1, le=200, description="Maximum results per page")
    offset: int = Field(0, ge=0, description="Number of results to skip")


# Response schemas for operations
class OperationResponse(BaseModel):
    """Generic operation response schema."""

    success: bool
    message: Optional[str] = None
    data: Optional[Dict[str, Any]] = Field(default_factory=dict)


class CreateAssessmentResponse(BaseModel):
    """Create assessment response schema."""

    success: bool
    assessment_id: str
    title: str
    status: str
    security_level: str
    created_at: str


class UpdateAnswerResponse(BaseModel):
    """Update answer response schema."""

    success: bool
    answer_id: str
    assessment_id: str
    control_id: str
    updated_at: str
    recalculation: Dict[str, Any] = Field(default_factory=dict)


class StatusTransitionResponse(BaseModel):
    """Status transition response schema."""

    success: bool
    assessment_id: str
    old_status: str
    new_status: str
    updated_at: str
    compliance_status: Optional[str] = None
    compliance_percentage: Optional[float] = None


class ErrorResponse(BaseModel):
    """Error response schema."""

    detail: str = Field(description="Error message")
    error_code: Optional[str] = Field(
        None, description="Error code for programmatic handling"
    )
    validation_errors: Optional[List[str]] = Field(
        None, description="Validation error details"
    )


# ============================================================================
# V2 SCHEMAS WITH SUBMEASURE CONTEXT SUPPORT
# (Merged from assessment_v2.py)
# ============================================================================

class UpdateAnswerRequestV2(BaseModel):
    """Update answer request with submeasure context."""
    
    control_id: str = Field(description="ID of the control being answered")
    submeasure_id: str = Field(description="ID of the submeasure context for this answer")
    documentation_score: Optional[int] = Field(None, ge=1, le=5)
    implementation_score: Optional[int] = Field(None, ge=1, le=5)
    comments: Optional[str] = Field(None, max_length=2000)
    evidence_files: Optional[List[str]] = Field(default_factory=list)
    confidence_level: Optional[int] = Field(None, ge=1, le=5)

    model_config = {
        "json_schema_extra": {
            "example": {
                "control_id": "123e4567-e89b-12d3-a456-426614174000",
                "submeasure_id": "223e4567-e89b-12d3-a456-426614174000",
                "documentation_score": 4,
                "implementation_score": 3,
                "comments": "Policy implemented but needs updates",
                "evidence_files": ["policy.pdf", "procedure.docx"],
                "confidence_level": 4
            }
        }
    }


class BatchUpdateAnswersRequestV2(BaseModel):
    """Batch update answers request with submeasure context."""
    
    answers: List[UpdateAnswerRequestV2] = Field(min_length=1)


class ControlInQuestionnaireResponseV2(BaseModel):
    """Control information within questionnaire with submeasure context."""
    
    # Control identification
    id: str = Field(description="Control UUID")
    code: str = Field(description="Control code (e.g., POL-001)")
    name_hr: str = Field(description="Control name in Croatian")
    description_hr: Optional[str] = Field(None, description="Control description in Croatian")
    
    # Context information
    submeasure_id: str = Field(description="Submeasure UUID this control instance belongs to")
    instance_id: str = Field(description="Unique identifier combining control_id:submeasure_id")
    
    # Requirements
    order_index: int
    is_mandatory: bool = Field(description="Whether control is mandatory at this security level")
    is_applicable: bool = Field(description="Whether control applies to this submeasure")
    minimum_score: Optional[float] = Field(None, description="Minimum required score for this control in this context")
    
    # Current answer (if exists)
    documentation_score: Optional[int] = Field(None, ge=1, le=5)
    implementation_score: Optional[int] = Field(None, ge=1, le=5)
    average_score: Optional[float] = Field(None, description="Calculated average of both scores")
    meets_minimum: Optional[bool] = Field(None, description="Whether current score meets minimum requirement")
    
    # Answer metadata
    comments: Optional[str] = None
    evidence_files: Optional[List[str]] = Field(default_factory=list)
    answered_at: Optional[datetime] = None
    answered_by: Optional[str] = None
    
    model_config = {
        "json_schema_extra": {
            "example": {
                "id": "123e4567-e89b-12d3-a456-426614174000",
                "code": "POL-001",
                "name_hr": "Politika informacijske sigurnosti",
                "submeasure_id": "223e4567-e89b-12d3-a456-426614174000",
                "instance_id": "123e4567-e89b-12d3-a456-426614174000:223e4567-e89b-12d3-a456-426614174000",
                "order_index": 1,
                "is_mandatory": True,
                "is_applicable": True,
                "minimum_score": 3.5,
                "documentation_score": 4,
                "implementation_score": 3,
                "average_score": 3.5,
                "meets_minimum": True
            }
        }
    }


class SubmeasureInQuestionnaireResponseV2(BaseModel):
    """Submeasure with controls maintaining context."""
    
    id: str
    code: str
    number: str = Field(description="Submeasure number (e.g., 1.1)")
    name_hr: str
    description_hr: Optional[str] = None
    order_index: int
    
    # Controls within this submeasure context
    controls: List[ControlInQuestionnaireResponseV2]
    
    # Aggregate scoring
    total_controls: int
    answered_controls: int
    mandatory_controls: int
    mandatory_answered: int
    controls_meeting_minimum: int = Field(description="Number of controls meeting their minimum scores")
    mandatory_meeting_minimum: int = Field(description="Number of mandatory controls meeting minimum")
    
    # Calculated scores
    documentation_avg: Optional[float] = None
    implementation_avg: Optional[float] = None
    overall_score: Optional[float] = None
    
    # Compliance status
    is_compliant: bool = Field(description="All mandatory controls answered and meeting minimum")
    compliance_issues: List[str] = Field(default_factory=list, description="List of compliance issues")


class MeasureInQuestionnaireResponseV2(BaseModel):
    """Measure containing submeasures with proper context."""
    
    id: str
    code: str
    name_hr: str
    description_hr: Optional[str] = None
    order_index: int
    
    submeasures: List[SubmeasureInQuestionnaireResponseV2]
    
    # Aggregate fields
    total_controls: int
    answered_controls: int
    mandatory_controls: int
    mandatory_answered: int
    overall_score: Optional[float] = None
    is_compliant: bool


class QuestionnaireResponseV2(BaseModel):
    """Complete questionnaire structure with submeasure context support."""
    
    assessment_id: str
    security_level: str
    version_id: Optional[str] = None
    measures: List[MeasureInQuestionnaireResponseV2]
    
    # Statistics
    statistics: dict = Field(description="Overall questionnaire statistics")
    
    # Metadata
    generated_at: datetime
    supports_multi_context: bool = Field(
        default=True, 
        description="Indicates this questionnaire supports controls in multiple submeasures"
    )


# Frontend-specific types for better TypeScript generation
class AnswerSubmissionV2(BaseModel):
    """Single answer submission with context."""
    
    control_id: str
    submeasure_id: str
    documentation_score: Optional[int] = None
    implementation_score: Optional[int] = None
    comments: Optional[str] = None
    
    model_config = {
        "from_attributes": True,
        "json_schema_extra": {
            "example": {
                "control_id": "123e4567-e89b-12d3-a456-426614174000",
                "submeasure_id": "223e4567-e89b-12d3-a456-426614174000",
                "documentation_score": 4,
                "implementation_score": 3,
                "comments": "Recently updated policy"
            }
        }
    }
